wfe2/wfe.go

```diff
@@ -2286,23 +2283,28 @@ func (wfe *WebFrontEndImpl) NewOrder(
 			return
 		}
 		if ident.Value == "" {
-			wfe.sendError(response, logEvent, probs.Malformed("NewOrder request included empty domain name"), nil)
+			wfe.sendError(response, logEvent, probs.Malformed("NewOrder request included empty identifier"), nil)
 			return
 		}
-		names[i] = ident.Value
 	}
+	idents = identifier.Normalize(idents)
+	logEvent.Identifiers = idents
 
-	names = core.UniqueLowerNames(names)
-	err = policy.WellFormedIdentifiers(identifier.NewDNSSlice(names))
+	err = policy.WellFormedIdentifiers(idents)
 	if err != nil {
 		wfe.sendError(response, logEvent, web.ProblemDetailsForError(err, "Invalid identifiers requested"), nil)
 		return
 	}
 
-	logEvent.DNSNames = names
+	pbIdents := identifier.ToProtoSlice(idents)
+	// TODO(#7311): Handle non-DNS identifiers.
+	names, err := identifier.ToDNSSlice(identifier.FromProtoSlice(pbIdents))
+	if err != nil {
+		wfe.sendError(response, logEvent, web.ProblemDetailsForError(err, "non-DNS identifiers requested"), nil)
+	}
```

Comment:
This dance a) isn't a very clear way to detect if there are any non-DNS identifiers, and b) isn't necessary because it's duplicating a check that's already above on line 2278.

Reply:
I'm cleaning up this area and inlining `pbIdents`, but the (redundant) error check seems like it's still needed, unless we want to drop an (unlikely!) error from `ToDNSSlice` on the floor? I thought about cleverly making the `ToDNSSlice` call part of the first check, but that would squelch the detail about which identifier was the problem, which seems like a useful part of the error message.

Reply:
Getting rid of pbIdents here makes this sufficiently clear and correct, I think this is fine now.